src/qt3utils/datagenerators/spectrometers/spectrometer.py

```diff
@@ -0,0 +1,230 @@
+import abc
+import inspect
+import logging
+
+from typing import Any, Callable, Dict, List, Literal, Tuple
+
+import numpy as np
+
+
+def get_method_argument_names(method: Callable) -> List[str]:
+    """
+    Returns a method's argument names.
+    """
+    return list(inspect.signature(method).parameters.keys())
+
+
+def filter_only_valid_kwargs_for_method(method: Callable, all_kwargs: dict) -> Dict[str, Any]:
+    """
+    Filters out kwargs that are not valid for the given method.
+    """
+    return {k: v for k, v in all_kwargs.items() if k in get_method_argument_names(method)}
+
+
+class Spectrometer(abc.ABC):
+
+    DEVICE_NAME: str = ''
+    """ The name of the device used for logging purposes. """
+
+    ACQUISITION_MODES: set[str] = {'single', 'step-and-glue', 'kinetic series', 'accumulation'}
+    """ The supported acquisition modes. """
+
+    def __init__(self):
+        logging.basicConfig(level=logging.INFO)
+        self.logger = logging.getLogger(self.DEVICE_NAME if self.DEVICE_NAME != '' else self.__class__.__name__)
+
+    @abc.abstractmethod
+    def open(self) -> None:
+        """
+        Initializes the connection to devices.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def close(self) -> None:
+        """
+        Terminates connection to devices.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    @property
+    def grating_list(self) -> List[str]:
+        """
+        A list of all installed gratings.
+        """
+        return []
+
+    @abc.abstractmethod
+    @property
+    def current_grating(self) -> str:
+        """
+        Current spectrometer grating.
+        """
+        return ''
+
+    @abc.abstractmethod
+    @current_grating.setter
+    def current_grating(self, value: str) -> None:
+        """
+        Set the spectrometer grating.
+        """
+        pass
+
+    @abc.abstractmethod
+    @property
+    def center_wavelength(self) -> float:
+        """
+        The grating center wavelength.
+        """
+        return np.nan
```

Comment:
Why is it np.nan?

Reply:
since these are abstract methods, they will be anyway. I can raise a `NotImplementedError` instead